### PR TITLE
Add timeout for stoppable post request

### LINKED ISSUE #1012: TODO: Add Timeout for Stoppable POST requests
It has been observed that sometime Helix REST server is trying to get instance health check from the outside sources (i.e. storage node) and SN does not respond at all. Therefore, a timeout for these requests needs to be added.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #1012 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In this PR, http timeout has been added for the POST requests
which is mainly used for REST's stoppable check.

### Tests
- [x] The following tests are written for this issue:
TestCustomRestClient.testGetPartitionStoppableCheck_when_timeout

- [x] The following is the result of the "mvn test" command on the appropriate module:

Helix-Core
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestStateTransitionTimeout.testStateTransitionTimeOut:170 expected:<true> but was:<false>
[ERROR]   TestUpdateWorkflow.testUpdateRunningQueue:82 expected:<true> but was:<false>
[ERROR]   TestWorkflowTermination.testWorkflowPausedTimeout:170->verifyWorkflowCleanup:257 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 1146, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:22 h
[INFO] Finished at: 2020-05-14T21:41:45-07:00
[INFO] ------------------------------------------------------------------------
```

The failed tests have passed when I ran them individually.
```
mvn test -Dtest="TestStateTransitionTimeout,TestUpdateWorkflow,TestWorkflowTermination"
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 62.382 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:07 min
[INFO] Finished at: 2020-05-15T11:29:16-07:00
[INFO] ------------------------------------------------------------------------
```
Helix-REST:
```
[INFO] Tests run: 160, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 41.568 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 160, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  47.282 s
[INFO] Finished at: 2020-05-15T11:46:34-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java [hunk 1]
```diff
@@ -78,4 +78,7 @@
   // System Property Metadata Store Directory Server endpoint key
   public static final String MSDS_SERVER_ENDPOINT_KEY =
       MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY;
+
+  // System property for request timeout
+  public static final String HTTP_REQUEST_TIMEOUT = "http.request.timeout";
```
Comment (narendly) [inline helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:83]: Is this strictly for Helix REST? In that case, should we try to put this in a PropertyKey constant file in the helix-rest module?

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java [hunk 2]
```diff
@@ -32,10 +36,13 @@
  */
 public class CustomRestClientFactory {
   private static final Logger LOG = LoggerFactory.getLogger(CustomRestClientFactory.class);
+  private static final int DEFAULT_HTTP_REQUEST_TIMEOUT = 60 * 1000;
```
Comment (narendly) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:39]: For timeout values, we usually use `long` type.

`60 * 1000L;`

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java [hunk 3]
```diff
@@ -44,14 +51,17 @@ public static CustomRestClient get() {
         if (INSTANCE == null) {
           try {
             HttpClient httpClient;
+            RequestConfig config = RequestConfig.custom().setConnectTimeout(_httpRequestTimeout)
+                .setConnectionRequestTimeout(_httpRequestTimeout)
+                .setSocketTimeout(_httpRequestTimeout).build();
```
Comment (narendly) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:56]: Do we want to set the socket and connect timeout as well? Those two seem unrelated to the problem you're solving. I am not saying we shouldn't necessarily, but we should understand the implications of setting such configs.
Also it would be good to add a line of comment describing what you're doing here.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 4]
```diff
@@ -145,6 +150,19 @@ public void testPostRequestFormat() throws IOException {
...
+  public void testPostRequestSmallTimeout() throws IOException {
+    // Set 1 ms to cause timeout for requests
+    System.setProperty(SystemPropertyKeys.HTTP_REQUEST_TIMEOUT, "1");
+    // a popular echo server that echos all the inputs
+    final String echoServer = "http://httpbin.org/post";
+    CustomRestClient _customRestClient = MockCustomRestClientFactory.get();
+    _customRestClient.getInstanceStoppableCheck(echoServer, Collections.emptyMap());
+
+    // Reset the HTTP_REQUEST_TIMEOUT property back to its original value
+    System.setProperty(SystemPropertyKeys.HTTP_REQUEST_TIMEOUT, "60000");
```
Comment (narendly) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:163]: :"back to its original value" -> where is the original value set?

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 5]
```diff
@@ -161,4 +179,35 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
...
   }
+
+  /**
+   * This MockCustomRestClientFactory is necessary to have for testing because once an INSTANCE is
+   * initialized in CustomRestClientFactory while running "mvn test" , it will no re-initialize the
+   * INSTANCE with new HelixProperty. Hence this class makes sure that new CustomRestClient will be
+   * created with the timeout set to the new value.
+   */
+  private static class MockCustomRestClientFactory extends CustomRestClientFactory {
+    private static final int DEFAULT_HTTP_REQUEST_TIMEOUT = 60 * 1000;
```
Comment (narendly) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:190]: Use long

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 6]
```diff
@@ -161,4 +179,35 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
       return new ObjectMapper().readTree(_jsonResponse);
     }
   }
+
+  /**
+   * This MockCustomRestClientFactory is necessary to have for testing because once an INSTANCE is
+   * initialized in CustomRestClientFactory while running "mvn test" , it will no re-initialize the
```
Comment (narendly) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:185]: Please fix typos.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java (hunk 1 above)
Comment (alirezazamani) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:83]: Done. Added a file called RestSystemPropertyKeys and added this key to that file.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java (hunk 2 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:39]: I noticed that as well. But since the setSocketTimeout and setConnectionRequestTimeout methods are getting int as input, I prefer to keep them as int. If not we need to have additional type casting.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java (hunk 3 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:56]: I think we prefer to have a timeout for any reason and not be blocked to query SN. That is why I used both of them as same value. Also I noticed that we do same for other classes such as HttpRoutingDataReader. I do not have strong preferences about this. I feel we should reduce the amount of configs as much as possible.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 5 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:190]: Please see my answer above. If you strongly feel we should do long, I can do it. But I think this way the code will be cleaner.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java (hunk 2 above)
Comment (narendly) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:39]: Interesting. In that case, could we leave a comment? That way, the intention is clear, and as the writer of this PR, you can justify using an int.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 5 above)
Comment (narendly) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:190]: in that case, using an int is fine. But can we avoid creating yet another constant here? I believe we already define it elsewhere.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java [hunk 7]
```diff
@@ -32,10 +36,13 @@
  */
 public class CustomRestClientFactory {
   private static final Logger LOG = LoggerFactory.getLogger(CustomRestClientFactory.class);
+  private static final int DEFAULT_HTTP_REQUEST_TIMEOUT = 60 * 1000;
+  private static final int _httpRequestTimeout = HelixUtil.getSystemPropertyAsInt(
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:40]: I combined the Default and getSystemProperty. Thanks.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 8]
```diff
@@ -130,6 +131,11 @@ protected HttpResponse post(String url, Map<String, String> payloads) throws IOE
       LOG.info("Executing request: {}, headers: {}, entity: {}", postRequest.getRequestLine(),
           postRequest.getAllHeaders(), postRequest.getEntity());
       return _httpClient.execute(postRequest);
+    } catch (ConnectTimeoutException e) {
+      LOG.error(
+          "Failed to perform customized health check due to ConnectTimeoutException for endpoint {}.",
+          url, e);
+      throw e;
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:138]: Removed the log and exception as it will be catched (IOException) by upper methods. Thanks.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 9]
```diff
@@ -161,4 +179,35 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
...
+   * INSTANCE with new HelixProperty. Hence this class makes sure that new CustomRestClient will be
+   * created with the timeout set to the new value.
+   */
+  private static class MockCustomRestClientFactory extends CustomRestClientFactory {
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:189]: Very good suggestion. Addressed. Thanks.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java [hunk 10]
```diff
@@ -32,10 +36,13 @@
...
+      SystemPropertyKeys.HTTP_REQUEST_TIMEOUT, DEFAULT_HTTP_REQUEST_TIMEOUT);
 
   private static CustomRestClient INSTANCE = null;
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:43]: Yeah it can be in another PR. Thanks for mentioning it here.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 6 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:185]: Removed. Thanks. I realize this class can be removed and not needed.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 4 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:163]: Fixed. Thanks.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 11]
```diff
@@ -145,6 +150,19 @@ public void testPostRequestFormat() throws IOException {
...
   }
 
+  @Test(expectedExceptions = ConnectTimeoutException.class)
+  public void testPostRequestSmallTimeout() throws IOException {
+    // Set 1 ms to cause timeout for requests
+    System.setProperty(SystemPropertyKeys.HTTP_REQUEST_TIMEOUT, "1");
+    // a popular echo server that echos all the inputs
+    final String echoServer = "http://httpbin.org/post";
+    CustomRestClient _customRestClient = MockCustomRestClientFactory.get();
+    _customRestClient.getInstanceStoppableCheck(echoServer, Collections.emptyMap());
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:160]: Basically wanted to check getInstanceStoppableCheck. I will change the Test's name. And I will change it to try catch.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 12]
```diff
@@ -145,6 +150,19 @@ public void testPostRequestFormat() throws IOException {
     Assert.assertEquals(json.get("data").asText(), "{}");
   }
 
+  @Test(expectedExceptions = ConnectTimeoutException.class)
+  public void testPostRequestSmallTimeout() throws IOException {
+    // Set 1 ms to cause timeout for requests
+    System.setProperty(SystemPropertyKeys.HTTP_REQUEST_TIMEOUT, "1");
+    // a popular echo server that echos all the inputs
+    final String echoServer = "http://httpbin.org/post";
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:158]: I changed the real server. Now, I am using non-routable internal address which causes the ConnectTimeoutException for http requests.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 5 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:190]: I added comments about it. I removed the default constant and directly uses the value in getSystemPropertyAsInt to avoid creating new constants. I found some other default values in for http request for MSDS related code. However, I am not sure if that is a good idea to have one default for everything. Anyway, please have a look and see if this makes sense.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java (hunk 2 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:39]: Sure. Added a comments.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:162]: I just realize that 127.0.0.0 gets timeout exception in mac only and not the Linux. So I changed it back to the external source for now. I added todo for this series of test to add mock rest server sometime in the future.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java [hunk 13]
```diff
@@ -32,26 +37,32 @@
...
 
   public static CustomRestClient get() {
     if (INSTANCE == null) {
       synchronized (CustomRestClientFactory.class) {
         if (INSTANCE == null) {
           try {
+            // Here int has been used for timeout value because setConnectTimeout,
+            // setConnectionRequestTimeout and setSocketTimeout are getting int as input
+            final int httpRequestTimeout =
+                HelixUtil.getSystemPropertyAsInt(RestSystemPropertyKeys.HTTP_TIMEOUT_MS, 60 * 1000);
```
Comment (dasahcc) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java:53]: Let's put this number to some constant file or at least to be a variable

Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientFactory.java]: Sure. Moved to constant file.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 14]
```diff
@@ -145,6 +148,35 @@ public void testPostRequestFormat() throws IOException {
...
+    CustomRestClient _customRestClient = CustomRestClientFactory.get();
+    boolean timeoutExceptionHappened = false;
+    try {
+      _customRestClient.getInstanceStoppableCheck(echoServer, Collections.emptyMap());
+    } catch (ConnectTimeoutException e) {
+      // Since the timeout is so small, we are expecting to get ConnectTimeoutException
+      timeoutExceptionHappened = true;
+    }
+    // Reset the HTTP_REQUEST_TIMEOUT property back to the default value
+    System.setProperty(RestSystemPropertyKeys.HTTP_TIMEOUT_MS, "60000");
```
Comment (pkuwm) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:172]: Why is this reset to a hard coded timeout? What if the timeout is changed, you setting it here 60000 would break the original value. The correct way should be to backup the original property and set it back here.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 15]
```diff
@@ -145,6 +148,35 @@ public void testPostRequestFormat() throws IOException {
...
+    System.setProperty(RestSystemPropertyKeys.HTTP_TIMEOUT_MS, "1");
+    // a popular echo server that echos all the inputs
+    // TODO: add a mock rest server
+    final String echoServer = "http://httpbin.org/post";
+    CustomRestClient _customRestClient = CustomRestClientFactory.get();
+    boolean timeoutExceptionHappened = false;
+    try {
+      _customRestClient.getInstanceStoppableCheck(echoServer, Collections.emptyMap());
+    } catch (ConnectTimeoutException e) {
+      // Since the timeout is so small, we are expecting to get ConnectTimeoutException
```
Comment (pkuwm) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:168]: I am not sure if I am convinced by this. What if it is fast enough to complete the request in 1 ms? It may not happen, but theoretically it could. This is relying on possibility of timeout..

File: helix-rest/src/main/java/org/apache/helix/rest/common/RestSystemPropertyKeys.java [hunk 16]
```diff
@@ -0,0 +1,6 @@
+package org.apache.helix.rest.common;
+
+public class RestSystemPropertyKeys {
+  // System property for request timeout
+  public static final String HTTP_TIMEOUT_MS = "http.timeout.ms";
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/common/RestSystemPropertyKeys.java:5]: Is this changed? Like I said in previous thread, The property name is too generic. If this is set, it indicates that all http in helix should be using the config. However this is not the case. Should be narrowed down to helix rest or even stoppable.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 14 above)
Comment (pkuwm) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:172]: Should use try finally to recover the state. Otherwise if this test throws exception before recovering, later tests may be failing.

File: helix-rest/src/main/java/org/apache/helix/rest/common/RestSystemPropertyKeys.java (hunk 16 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/common/RestSystemPropertyKeys.java]: Changed it to REST_HTTP_TIMEOUT_MS.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 15 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java]: Changed the test completely. It is using mock now.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java (hunk 14 above)
Comment (alirezazamani) [reply helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java]: Removed this test.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @dasahcc 

Final commit message:

Add timeout for stoppable post request

In this commit, a http timeout has been added for the POST requests
which is mainly used for REST's stoppable check.

File: helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java [hunk 17]
```diff
@@ -145,6 +146,28 @@ public void testPostRequestFormat() throws IOException {
     Assert.assertEquals(json.get("data").asText(), "{}");
   }
 
+  @Test
+  public void testGetPartitionStoppableCheck_when_timeout() throws IOException {
```
Comment (pkuwm) [inline helix-rest/src/test/java/org/apache/helix/rest/client/TestCustomRestClient.java:150]: This mix of styling (snake_case and camelCase) isn't java style

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 18]
```diff
@@ -33,6 +33,7 @@
 import org.apache.http.client.ClientProtocolException;
 import org.apache.http.client.HttpClient;
 import org.apache.http.client.methods.HttpPost;
+import org.apache.http.conn.ConnectTimeoutException;
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:36]: Please remove this unused import
